soccer/planning/planner/SettlePlanner.cpp

```diff
@@ -0,0 +1,633 @@
+void SettlePlanner::processStateTransition(BallState ball,
+                                           RobotInstant* startInstant,
+                                           double angle,
+                                           Geometry2d::Point deltaPos) {
+    // State transitions
+    // Intercept -> Dampen, PrevPath and almost at the end of the path
+    // Dampen -> Complete, PrevPath and almost slowed down to 0?
+    if (!previous.empty() && startInstant->stamp > previous.begin_time() &&
+        startInstant->stamp <= previous.end_time()) {
+        Geometry2d::Line ballMovementLine(ball.position,
+                                          ball.position + averageBallVel);
+
+        Trajectory pathSoFar =
+            previous.subTrajectory(previous.begin_time(), startInstant->stamp);
+        double botDistToBallMovementLine =
+            ballMovementLine.distTo(pathSoFar.last().position() - deltaPos);
+
+        // Intercept -> Dampen
+        //  Almost intersecting the ball path and
+        //  Almost at end of the target path or
+        //  Already in line with the ball
+        //
+        // TODO(Kyle): Check ball sense?
+
+        // Within X seconds of the end of path
+        bool inlineWithBall =
+            botDistToBallMovementLine < cos(angle) * Robot_MouthWidth / 2;
+        bool inFrontOfBall =
+            averageBallVel.angleBetween(startInstant->position() -
+                                        ball.position) < 3.14 / 2;
+
+        if (inFrontOfBall && inlineWithBall &&
+            currentState == SettlePlannerStates::Intercept) {
+            // Start the next section of the path from the end of our current
+            // path
+            *startInstant = pathSoFar.last();
+            currentState = SettlePlannerStates::Dampen;
+        }
+    }
+}
+
+Trajectory SettlePlanner::intercept(
+    const PlanRequest& planRequest, RobotInstant startInstant,
+    const Geometry2d::ShapeSet& staticObstacles,
+    const std::vector<DynamicObstacle>& dynamicObstacles,
+    Geometry2d::Point deltaPos, Geometry2d::Point facePos) {
+    BallState ball = planRequest.world_state->ball;
+
+    // First, if the previous trajectory is still valid, try to use that.
+    if (!previous.empty() && previous.CheckTime(startInstant.stamp) &&
+        !TrajectoryHitsStatic(previous, staticObstacles, startInstant.stamp,
+                              nullptr) &&
+        !TrajectoryHitsDynamic(previous, dynamicObstacles, startInstant.stamp,
+                               nullptr, nullptr)) {
+        RJ::Seconds timeRemaining = previous.end_time() - startInstant.stamp;
+        Point pathAtEnd = previous.last().position();
+
+        Point nearPoint;
+        RJ::Seconds timeToIntersect =
+            ball.query_seconds_near(pathAtEnd, &nearPoint);
+        if (timeRemaining <
+                timeToIntersect + RJ::Seconds(*_interceptBufferTime) &&
+            nearPoint.distTo(pathAtEnd) < Replanner::goalPosChangeThreshold()) {
```

Comment:
This will need to be tested, but you may run into problems with the `goalPosChangeThreshold` being larger than the `mouthWidth/2` so the reused path won't actually get you close enough to the ball path to actually catch it

Reply:
This is especially problematic since the "shortcut" stuff is cut out with this early return. The ball velocity is surprisingly noisy

Reply:
Yeah, I'll remove this. It's no longer really an issue now that replanning is working a lot better.